drivers/wifi/infineon/cyw43xxx_drv.c

```diff
@@ -0,0 +1,716 @@
+			LOG_ERR("Error Unable to bring up the whd secondary interface");
+			k_sem_give(&data->sema_common);
+			return -EAGAIN;
+		}
+		data->second_interface_init = 1;
+	}
+
+	if (data->is_ap_up) {
+		LOG_ERR("Already AP is on - first disconnect");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	ssid.length = params->ssid_length;
+	memcpy((void *)ssid.value, (void *)params->ssid, ssid.length);
+
+	if (((params->channel > 0) && (params->channel < 12)) ||
+	    ((params->channel > 35) && (params->channel < 166))) {
+		channel = params->channel;
+	} else {
+		channel = 1;
+	}
+
+	if (params->psk_length == 0) {
+		security = WHD_SECURITY_OPEN;
+	} else {
+		security = WHD_SECURITY_WPA2_AES_PSK;
+	}
+
+	if (whd_wifi_init_ap(cyw43xxx_ap_if, &ssid, security, (const uint8_t *)params->psk,
+			     params->psk_length, channel) != 0) {
+		LOG_ERR("Failed to init whd ap interface");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	if (whd_wifi_start_ap(cyw43xxx_ap_if) != 0) {
+		LOG_ERR("Failed to start whd ap interface");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	/* set event handler */
+	if (whd_management_set_event_handler(cyw43xxx_ap_if, ap_link_events, ap_link_events_handler,
+					     NULL, &ap_event_handler_index) != 0) {
+		whd_wifi_stop_ap(cyw43xxx_ap_if);
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	data->is_ap_up = 1;
+
+	k_sem_give(&data->sema_common);
+	return 0;
+}
+
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+static int cyw43xxx_mgmt_wifi_stats(const struct device *dev, struct net_stats_wifi *stats)
+{
+	struct cyw43xxx_wifi_runtime *data = dev->data;
+
+	stats->bytes.received = data->stats.bytes.received;
+	stats->bytes.sent = data->stats.bytes.sent;
+	stats->pkts.rx = data->stats.pkts.rx;
+	stats->pkts.tx = data->stats.pkts.tx;
+	stats->errors.rx = data->stats.errors.rx;
+	stats->errors.tx = data->stats.errors.tx;
+	stats->broadcast.rx = data->stats.broadcast.rx;
+	stats->broadcast.tx = data->stats.broadcast.tx;
+	stats->multicast.rx = data->stats.multicast.rx;
+	stats->multicast.tx = data->stats.multicast.tx;
+	stats->sta_mgmt.beacons_rx = data->stats.sta_mgmt.beacons_rx;
+	stats->sta_mgmt.beacons_miss = data->stats.sta_mgmt.beacons_miss;
+
+	return 0;
+}
+#endif
+
+static int cyw43xxx_mgmt_ap_disable(const struct device *dev)
+{
+	int ret = 0;
+	struct cyw43xxx_wifi_runtime *data = (struct cyw43xxx_wifi_runtime *)dev->data;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(CYW43XXX_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	whd_wifi_deregister_event_handler(cyw43xxx_ap_if, ap_event_handler_index);
+	whd_wifi_stop_ap(cyw43xxx_ap_if);
+	data->is_ap_up = 0;
+
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+static int cyw43xxx_init(const struct device *dev)
+{
+	cy_rslt_t ret = CY_RSLT_SUCCESS;
+	struct cyw43xxx_wifi_runtime *data = (struct cyw43xxx_wifi_runtime *)dev->data;
+
+	cyhal_gpio_t sdio_cmd =
+		DT_GET_CYHAL_GPIO_FROM_DT_GPIOS(DT_WIFI_SDIO_NODE, sdio_cmd_gpios);
```

Comment:
Hey, so just to understand, this means that this driver will only ever work when paired with `infineon_cat1` devices, right?

Reply:
what is preventing from using `CONFIG_SDIO_STACK`?

Reply:
Yes, current version of this is designed to use Infineon cat1 devices, which support cat1hal.
 
_"what is preventing from using CONFIG_SDIO_STACK?"_
we have not support SDIO driver yet. wifi-host-driver (whd) uses cyhal for sdio/gpio functionality, so i added DT_GET_CYHAL_GPIO_FROM_DT_GPIOS macros to retrieve cyhal_gpio object (port#/pin#) from  DT_GPIOS which we provide from board.dts or board overlay..

_"'I'm  also surprised by DT_ namespace being used out of DT API... "_
Let me know if it make any problem, maybe we can rename it.

Reply:
> Yes, current version of this is designed to use Infineon cat1 devices, which support cat1hal.
> 
> _"what is preventing from using CONFIG_SDIO_STACK?"_ we have not support SDIO driver yet. wifi-host-driver (whd) uses cyhal for sdio/gpio functionality, so i added DT_GET_CYHAL_GPIO_FROM_DT_GPIOS macros to retrieve cyhal_gpio object (port#/pin#) from DT_GPIOS which we provide from board.dts or board overlay..
> 

Then what is the point of providing such kind of drivers upstream? Sounds like this should be kept in a vendor SDK until ready because this driver is unusable on all platforms (e.g. new arduino giga) except for Infineon SoCs.

Reply:
@npal-cy  and @gmarull, yes, your thoughts are correct. I would like to do this in three stages:
1. Get the driver as it is up now for existing customer accounts (hal_infineon is fine)
2. Work around the issues that prevent use of the driver with non-Infineon hosts via some simple patches to the Infineon Wi-Fi Driver
3. Work internally for a more robust, patch free solution that makes the Infineon Wi-Fi Driver easier to port to non-Infineon SDKs (regardless of host)

Thanks,

Ian

Reply:
Hi @gmarull, could you please confirm that we must move cyw43xxx wifi driver to hal_infineon?

Reply:
Hey folks, apologies for escalating this but I think it deserves its own discussion. I filed https://github.com/zephyrproject-rtos/zephyr/issues/61798 for the specific situation and flagged it for Architecture review, feel free to chime in on the issue, hope to have a discussion about it next week.

Reply:
SOF has a few:

```
./audio/sof/src/platform/posix/dma.c:DEVICE_DEFINE(pzdma0, "pzdma0", pzdma_init, NULL, &pzdma0_data, &pzdma0_cfg,
```

You probably just want some path for the CI to build it but I don't think we have a policy about it.

Reply:
@gmarull, we also have Zephyr/modules/hal_infineon folder. Is it a option to move cyw43xx wifi driver to this folder for now  (instead hal_infineon repo) ?

Reply:
I'd not allow such kinds of drivers in tree, regardless of their location. But I don't know who can decide that.

Reply:
I think the hal module would make more sense in this case, though it's not quite a hal

@carlescufi whta do you think?